### ISSUE TITLE
Transforming a directory of resources does not do anything

### ISSUE THREAD

**Issue by sebersole** [eclipse/transformer#148]
Attempting to transform a directory of resources does nothing.  No output is ever written.

Sample output:

```
Copyright (c) 2020 Contributors to the Eclipse Foundation
This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
which is available at https://www.apache.org/licenses/LICENSE-2.0.
SPDX-License-Identifier: (EPL-2.0 OR Apache-2.0)

org.eclipse.transformer.Transformer
  Version [ 0.2.0 ]

Log file [ null ]
[main] INFO Transformer - Input     [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-750697555/shadowMulti/real/build/resources/main ]
[main] INFO Transformer -           [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-750697555/shadowMulti/real/build/resources/main ]
[main] INFO Transformer - Output    [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-750697555/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources ] (Explicit)
[main] INFO Transformer -           [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-750697555/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources ]
[main] INFO Transformer - Output already exists [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-750697555/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources ]
```

I've tried manually deleting the target directory first as well.  That actually seems to perform the transformation, although no file(s) are written to the target directory:

```

Copyright (c) 2020 Contributors to the Eclipse Foundation
This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
which is available at https://www.apache.org/licenses/LICENSE-2.0.
SPDX-License-Identifier: (EPL-2.0 OR Apache-2.0)

org.eclipse.transformer.Transformer
  Version [ 0.2.0 ]

Log file [ null ]
[main] INFO Transformer - Input     [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/real/build/resources/main ]
[main] INFO Transformer -           [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/real/build/resources/main ]
[main] INFO Transformer - Output    [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources ] (Explicit)
[main] INFO Transformer -           [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources ]
[main] INFO Transformer - Skipping option [ RULES_SELECTIONS ]
[main] INFO Transformer - External [ RULES_RENAMES ] URL [ file:/home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/rules/jakarta-renames.properties ]
[main] INFO Transformer - External [ RULES_VERSIONS ] URL [ file:/home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/rules/jakarta-versions.properties ]
[main] INFO Transformer - Internal [ RULES_BUNDLES ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-bundles.properties ]
[main] INFO Transformer - External [ RULES_DIRECT ] URL [ file:/home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/rules/jakarta-direct.properties ]
[main] INFO Transformer - Internal [ RULES_MASTER_TEXT ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-text-master.properties ]
[main] INFO Transformer - Internal [ RULES_PER_CLASS_CONSTANT ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-per-class-constant-master.properties ]
[main] INFO Transformer - Package renames are in use
[main] INFO Transformer - Package versions will be updated
[main] INFO Transformer - Bundle identities will be updated
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.TransactionSynchronizationRegistryService.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-transactionSynchronizationRegistry.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.UserTransactionService.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ application-client.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ TransactionJavaColonHelper.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.RemoteTransactionControllerService.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-transactionManager.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ web.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.webcontainer.v40.dd.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-servletVersion.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ UTService.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ application.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ TransactionManager.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-transactionManager.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ ejb-jar.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ web-fragment.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ permissions.xml ] ] URL [ jar:file:/tmp/.gradle-test-kit-sebersole/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Text files will be updated
[main] INFO Transformer - Java direct string updates will be performed
[main] INFO Transformer - Per class constant mapping files are not enabled
[main] INFO Transformer - All resources will be selected
[main] INFO Transformer - Per class constant mapping files are not enabled
[main] INFO Transformer - Action selected for input [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/real/build/resources/main ]: Directory Action
[main] INFO Transformer - Input [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/real/build/resources/main ] as [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources ]: Unchanged
[main] INFO Transformer - [          All Resources ] [      1 ] Unselected [      0 ] Selected [      1 ]
[main] INFO Transformer - [            All Actions ] [      1 ]  Unchanged [      1 ]  Changed [      0 ]
```

In this second attempt, you can see the files are processed and supposedly transformed.  But afterwards the output directory `/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources` is empty.

**Comment by sebersole** [comment]
Bueller?

**Comment by sebersole** [comment]
Actually in my experience and testing transforming a directory is very inconsistent.  Sure would be awesome to get some kind of validation here... Is this project even active?

**Comment by bjhargrave** [comment]
Sorry. I have been busy with other work and have not had much time lately to put into this project. I hope to find some time soon to make some improvements.

I cannot offer much advice here other than you probably have found a bug.

**Comment by sebersole** [comment]
Hi BJ, thanks for the reply.

Its just frustrating because this tool is critical for anyone to to switch to Jakarta from javax EE.  I won't even consider it for Hibernate without this stuff being able to work.

**Comment by tbitonti** [comment]
Hi, re: writing to an empty directory.  There is this option which should enable that:
		OVERWRITE("o", "overwrite", "Overwrite", !OptionSettings.HAS_ARG, !OptionSettings.HAS_ARGS,
			!OptionSettings.IS_REQUIRED, OptionSettings.NO_GROUP),
That is "-o" or "--overwrite".

Is there a case of a non-empty directory failing to copy to an empty or initially non-existent output directory?  That would be a bug.  Is there a data set (zipped directory) which can be provided which can be used to recreate the problem?

Or, is there a problem where if none of the resources of the directory are selected to be processed, then none are copied?

**Comment by sebersole** [comment]
@tbitonti Its simply the target directory existing.  If it already exists, the transformer fails even if that target directory is empty.  There are files in the source.

**Comment by tbitonti** [comment]
@sebersole 

This output indicates that the input directory is empty ... the transformer has a count only of the root directory, and none for nested files:

[main] INFO Transformer - Input [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/real/build/resources/main ] as [ /home/sebersole/projects/playground/jakarta-transformer-plugin/build/tmp/testKit/testKit-1717152281/shadowMulti/shadow/build/generated/sources/jakarta/transform/real/main/resources ]: Unchanged
[main] INFO Transformer - [          All Resources ] [      1 ] Unselected [      0 ] Selected [      1 ]
[main] INFO Transformer - [            All Actions ] [      1 ]  Unchanged [      1 ]  Changed [      0 ]

Regardless of the behavior when the output already exists, if the input folder isn't actually empty, that is a bug (and separate from the behavior question).

Currently, the design is to disallow any overwrites.  This is a specific design point: The transformer does not handle the management of the input and output locations.  For simplicity, the output must not exist unless the overwrite option is used.

There are two specific behaviors which could be built in -- having the output be the same as the input, and creating a backup of the initial output, if it already exists.  Both of these behaviors were beyond the initial needs of the transformer, and seem better handled external to the transformer.  This sort of function was deliberately excluded from the transformer.

The cases of the input and output being a simple file, or of the input and output not matching, or where the output is a non-empty directory, are working as designed.

I can see having a special case for an empty output directory.  The overwrite option (-o, --overwrite) should provide this capability, but this might not be the best solution. If a special case is made for outputting to an empty directory, that might lead to cases of process failures when the directory was not empty.  On the other hand, adding -o would enable output to a non-empty directory, when you might really only want output to a non-empty directory.

**Comment by sebersole** [comment]
Gradle will always create the directory because the task defines it as any output.  Gradle creates these on the task's behalf when the task is to  execute.. a very nice feature.  Which means this behavior will be a "conflict" for any one using this from Gradle; unless they use/write a custom task that does not properly define inputs/outputs, which in turn will sabotage Gradle's up-to-date checking.  

Y'alls call of course...

**Comment by sebersole** [comment]
To be clear, I'm not arguing that the tool "works as designed".  I'm saying that the design is not great in this particular case

**Comment by tbitonti** [comment]
I wasn't aware of that Gradle behavior.  That's a good enough reason (for me) to add in the special case.  Would the output directory always be empty?  I'd prefer to only enable this case for output to an existing empty directory.

**Comment by sebersole** [comment]
Yes, this is just if the directory is empty

**Comment by bjhargrave** [comment]
I just fixed the issue with needing overwrite for an empty output directory. https://github.com/eclipse/transformer/pull/244

As for not transforming a directory of resources, there is a test which does this: https://github.com/eclipse/transformer/blob/3592a554285644845dbf3fe621933d1ccfc3efad/org.eclipse.transformer/src/test/java/transformer/test/TestTransformerMultiText.java#L186.

Perhaps this has been fixed in the latest 0.3.0-SNAPSHOT? Can you try?

**Comment by bjhargrave** [comment]
Closing. Please open a new issue. The latest build does not complain about empty output directory.
